login.go

```diff
@@ -0,0 +1,211 @@
+package conductoroneapi
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type ClientCredentials struct {
+	ClientID     string
+	ClientSecret string
+}
+
+type DeviceCodeResponse struct {
+	DeviceCode      string `json:"device_code"`
+	UserCode        string `json:"user_code"`
+	VerificationURI string `json:"verification_uri_complete"`
+	ExpiresIn       int64  `json:"expires_in"`
+	Interval        int64  `json:"interval"`
+}
+
+func LoginFlow(ctx context.Context, tenantName string, clientID string, personalClientCredentialDisplayName string) (*ClientCredentials, error) {
+	opts := []SDKOption{}
+	// If they pass a URL, use the whole URL
+	if strings.Contains(tenantName, ".") {
+		opts = append(opts, WithServerURL(tenantName))
+	} else {
+		opts = append(opts, WithTenantDomain(tenantName))
+	}
+	client := New(opts...)
+
+	codeResp, err := getDeviceCode(ctx, client, clientID)
+	if err != nil {
+		return nil, fmt.Errorf("error getting device code: %w", err)
+	}
+
+	tokenResp, err := doTokenRequest(ctx, client, clientID, codeResp)
+	if err != nil {
+		return nil, fmt.Errorf("error doing token request: %w", err)
+	}
+
+	clientCredential, err := doClientCredentialRequest(ctx, client, tokenResp, personalClientCredentialDisplayName)
+	if err != nil {
+		return nil, fmt.Errorf("error doing client credential request: %w", err)
+	}
+
+	return &ClientCredentials{
+		ClientID:     clientCredential.Client.ClientID,
+		ClientSecret: clientCredential.ClientSecret,
+	}, nil
+}
+
+func newReq(ctx context.Context, url, method string, reader io.Reader) (*http.Request, error) {
+	req, err := http.NewRequestWithContext(ctx, method, url, reader)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("User-Agent", "conductorone-sdk-go")
+	return req, nil
+}
+
+func getDeviceCode(ctx context.Context, client *ConductoroneAPI, clientID string) (*DeviceCodeResponse, error) {
+	httpClient := client.sdkConfiguration.DefaultClient
+
+	deviceCodeURL := "https://" + client.sdkConfiguration.ServerURL + "/auth/v1/device_authorization"
+	vals := url.Values{}
+	vals.Add("client_id", clientID)
+
+	req, err := newReq(ctx, "POST", deviceCodeURL, strings.NewReader(vals.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	data, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	codeResp := &DeviceCodeResponse{}
+	err = json.Unmarshal(data, codeResp)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("Open: %s\n", codeResp.VerificationURI)
```

Comment:
should we use webbrowser here?

i think at the very least, Login() should take a callback function that is OpenURL(string)?  then the caller can control the UX?

Reply:
webborwser :https://pkg.go.dev/github.com/toqueteos/webbrowser

Reply:
I like the callback idea over the web browser idea by default. Although we can use webbrowser in cone.